### ISSUE TITLE
coap_make_str_const() does not work in C++

### ISSUE THREAD

**Issue by obgm** [obgm/libcoap#233]
When trying to use `coap_make_str_const()` as shown in `examples/coap-server.c` in C++, g++ 8.1.0 emits the following error and multiple warnings:

```
am.cc: In function ‘void init_resources(coap_context_t*)’:
/home/bergmann/build/include/coap/str.h:76:24: warning: C++ designated initializers only available with -std=c++2a or -std=gnu++2a [-Wpedantic]
   (&(coap_str_const_t){.length = sizeof(string)-1,.s = (const uint8_t *)(string)})
                        ^
am.cc:106:33: note: in expansion of macro ‘coap_make_str_const’
   resource = coap_resource_init(coap_make_str_const(RESOURCE_NAME), 0);
                                 ^~~~~~~~~~~~~~~~~~~
/home/bergmann/build/include/coap/str.h:76:51: warning: C++ designated initializers only available with -std=c++2a or -std=gnu++2a [-Wpedantic]
   (&(coap_str_const_t){.length = sizeof(string)-1,.s = (const uint8_t *)(string)})
                                                   ^
am.cc:106:33: note: in expansion of macro ‘coap_make_str_const’
   resource = coap_resource_init(coap_make_str_const(RESOURCE_NAME), 0);
                                 ^~~~~~~~~~~~~~~~~~~
/home/bergmann/build/include/coap/str.h:76:81: warning: ISO C++ forbids compound-literals [-Wpedantic]
   (&(coap_str_const_t){.length = sizeof(string)-1,.s = (const uint8_t *)(string)})
                                                                                 ^
am.cc:106:33: note: in expansion of macro ‘coap_make_str_const’
   resource = coap_resource_init(coap_make_str_const(RESOURCE_NAME), 0);
                                 ^~~~~~~~~~~~~~~~~~~
/home/bergmann/build/include/coap/str.h:76:81: error: taking address of temporary [-fpermissive]
   (&(coap_str_const_t){.length = sizeof(string)-1,.s = (const uint8_t *)(string)})
                                                                                 ^
am.cc:106:33: note: in expansion of macro ‘coap_make_str_const’
   resource = coap_resource_init(coap_make_str_const(RESOURCE_NAME), 0);
```

**Comment by mrdeep1** [comment]
Does it work with (removing .X = ) ?
````
#define coap_make_str_const(string) \
  (&(coap_str_const_t){sizeof(string)-1,(const uint8_t *)(string)})
````

**Comment by obgm** [comment]
No:

```
am.cc: In function ‘void init_resources(coap_context_t*)’:
/home/bergmann/build/include/coap/str.h:100:66: warning: ISO C++ forbids compound-literals [-Wpedantic]
   (&(coap_str_const_t){sizeof(string)-1,(const uint8_t *)(string)})
```

(Even if the initialization was allowed, the main error—`"taking address of temporary [-fpermissive]"` —still remains.)

**Comment by mrdeep1** [comment]
OK - thanks for trying - need to put my thinking cap on to come up with an alternative.

**Comment by obgm** [comment]
It took me some time to completely understand the issue. Commit 15a45f6 now has a fix.

**Comment by fero29** [comment]
I have still problem with coap_make_str_const

I use  Commit 15a45f6
c++14

`
Coap_session.cpp:31:30: error: no matching function for call to ‘libcoap::CoAPStrConst::CoAPStrConst(<brace-enclosed initializer list>)’
     coap_str_const_t *ruri = coap_make_str_const("hello");
                              ^
/usr/local/include/coap2/str.h:95:10: note: candidate: libcoap::CoAPStrConst::CoAPStrConst()
   struct CoAPStrConst : coap_str_const_t {
          ^~~~~~~~~~~~
/usr/local/include/coap2/str.h:95:10: note:   candidate expects 0 arguments, 2 provided
/usr/local/include/coap2/str.h:95:10: note: candidate: constexpr libcoap::CoAPStrConst::CoAPStrConst(const libcoap::CoAPStrConst&)
/usr/local/include/coap2/str.h:95:10: note:   candidate expects 1 argument, 2 provided
/usr/local/include/coap2/str.h:95:10: note: candidate: constexpr libcoap::CoAPStrConst::CoAPStrConst(libcoap::CoAPStrConst&&)
/usr/local/include/coap2/str.h:95:10: note:   candidate expects 1 argument, 2 provided
`

**Comment by mrdeep1** [comment]
@fero29 This has been superseded by #384 (commit 5d32ffe92067ce6b38730007e9a5f141b0c0904b) which you need to test against.

Please confirm if you still have the issue with that version of code (which is in 4.2.1) (or preferably the latest develop branch).

**Comment by fero29** [comment]
ok now I use version release-4.2.1 but it makes problem.
error: ‘coap_session_t {aka struct coap_session_t}’ has no member named ‘remote_addr’

and when I use fix #350 27eec0c  It revert to version 4.2.0

**Comment by mrdeep1** [comment]
@fero29 It would appear that the application code that you are using needs to be updated to the latest version of coap_session_t structure.

`remote_addr` is now a part of addr_info in coap_session_t so can be referenced by `addr_info.remote` as a substitute.  This change was made as a part of a performance update (#378).

Fix #350 was done some time ago and the code (and code version) has moved on since then.

**Comment by fero29** [comment]
thank you
